bin/differential_tester_x86/LiftAndCompare.cpp

```diff
@@ -0,0 +1,541 @@
+
+  virtual llvm::Type *remapType(llvm::Type *SrcTy) override {
+    LOG(INFO) << "Attempting to remap: " << remill::LLVMThingToString(SrcTy);
+    if (auto it = this->tmap.find(SrcTy); it != this->tmap.end()) {
+      return it->second;
+    }
+
+    return SrcTy;
+  }
+};
+
+void CloneFunctionWithTypeMap(llvm::Function *NewFunc, llvm::Function *OldFunc,
+                              remill::TypeMap &tmap) {
+
+  remill::ValueMap vmap;
+  remill::MDMap md_map;
+  remill::CloneFunctionInto(OldFunc, NewFunc, vmap, tmap, md_map);
+}
+
+class DifferentialModuleBuilder {
+ public:
+  static DifferentialModuleBuilder
+  Create(remill::OSName os_name_1, remill::ArchName arch_name_1,
+         remill::OSName os_name_2, remill::ArchName arch_name_2) {
+    // it is expected that compatible arches share a semantics module.
+    std::unique_ptr<llvm::LLVMContext> context =
+        std::make_unique<llvm::LLVMContext>();
+    context->enableOpaquePointers();
+    auto tmp_arch = remill::Arch::Build(context.get(), os_name_1, arch_name_1);
+    std::shared_ptr<llvm::Module> semantics_module =
+        remill::LoadArchSemantics(tmp_arch.get());
+    tmp_arch->PrepareModule(semantics_module.get());
+    auto l1 =
+        test_runner::LiftingTester(semantics_module, os_name_1, arch_name_1);
+    auto l2 =
+        test_runner::LiftingTester(semantics_module, os_name_2, arch_name_2);
+    return DifferentialModuleBuilder(std::move(context),
+                                     std::move(semantics_module), std::move(l1),
+                                     std::move(l2));
+  }
+
+ private:
+  std::unique_ptr<llvm::LLVMContext> context;
+  std::shared_ptr<llvm::Module> semantics_module;
+
+  test_runner::LiftingTester l1;
+  test_runner::LiftingTester l2;
+
+  DifferentialModuleBuilder(std::unique_ptr<llvm::LLVMContext> context_,
+                            std::shared_ptr<llvm::Module> semantics_module_,
+
+                            test_runner::LiftingTester l1_,
+                            test_runner::LiftingTester l2_)
+      : context(std::move(context_)),
+        semantics_module(std::move(semantics_module_)),
+        l1(std::move(l1_)),
+        l2(std::move(l2_)) {}
+
+ public:
+  std::optional<DiffModule> build(std::string_view fname_f1,
+                                  std::string_view fname_f2,
+                                  std::string_view bytes, uint64_t address) {
+    auto module = std::make_unique<llvm::Module>("", *this->context);
+    auto maybe_f1 = this->l1.LiftInstructionFunction(fname_f1, bytes, address);
+    auto maybe_f2 = this->l2.LiftInstructionFunction(fname_f2, bytes, address);
+
+    if (!maybe_f1.has_value() || !maybe_f2.has_value()) {
+      return std::nullopt;
+    }
+
+    auto f1_and_name = *maybe_f1;
+    auto f2_and_name = *maybe_f2;
+
+    auto f1 = f1_and_name.first;
+    auto f2 = f2_and_name.first;
+    llvm::verifyFunction(*f1, &llvm::errs());
+
+    llvm::verifyFunction(*f2, &llvm::errs());
```

Comment:
Copy & paster

Reply:
I tried to refactor this to handle f1 and f2 agnostically ie. refactor everything to handle n lifters, n lifted instructions. There is too much code that depends on this being a compare between two functions at this time, will try to reduce code duplication without refactoring to n functions.